### ISSUE TITLE
Development request: speed-up findpeaks

### ISSUE THREAD

**Issue by calum-chamberlain** [eqcorrscan/EQcorrscan#159]
A request was made outside of github to speed-up find-peaks.

I think this could be done, and could likely be done with a C-routine.  I will have a play and see if I can get it to work.

Alongside this, a suggestion of testing difference in detections between single-channel thresholds and stacked and summed thresholds was made.  I think this would be worth experimenting with in a branch too.

**Comment by d-chambers** [comment]
If I get a few minutes this weekend (not 100% sure I will) I can take a look at this. 

I think you could do it all in numpy, getting rid of the python lists will go a long way. That would be worth trying before jumping into C. 

Can you ask the user to provide an example dataset? Even if we speed it up for the test cases it could be quite a different story for actual data.

**Comment by calum-chamberlain** [comment]
You make an excellent point - if you do get the chance this weekend that would be excellent! Definitely keen to keep it in Python if we can.

I will get in touch for a test-dataset.

**Comment by yifanyin** [comment]
Hi guys! Here comes the github outsider. The reason I suggested this is because I perform cross-correlation for a long period of time (several years) using few stations (2-5) so I end up running find_peak for lots of times. I tried to parallelize the find_peak in the old eqcorrscan version by the same way it parallelizes match_filter but discarded it since it upgraded to Python 3.x.

About the single-channel threshold. I intended to use it as a threshold in addition to the summed one. I got questioned that my detections are dominated by S-arrivals. Independent lower threshold might serve as a double-checker.

**Comment by calum-chamberlain** [comment]
Do you think weighting the cross-correlation would help? You could weight template channels that contain only the P-wave more heavily than those containing only S-wave?
The reason I ask is that that is something that William Frank does with his correlations, the C-routines are on github [here](https://github.com/beridel/fast_matched_filter). These are just C-routines with Python and Matlab wrappers, so they don't have the front end that EQcorrscan does, but you can now define a custom xcorr function in EQcorrscan so you could install their package and test against that? It would require a bit of tweaking in EQcorrscan mind (to pass weights through), but if it is worth it it could be done.

A quick note if you do look at their correlations - they don't normalise in the same way that we do, so the correlations will be slightly different - for every step of the correlations they don't remove the mean of the continuous data.

**Comment by d-chambers** [comment]
Hey @calum-chamberlain,
I tried a few things but, on the dataset included in the tests, I couldn't beat the current function. I may be able to make the current function a bit faster (~20% maybe) but it probably isn't worth it. Maybe a C implementation is in order? Or maybe @yifanyin just needs a way to run it concurrently?

**Comment by calum-chamberlain** [comment]
Thanks for having a go - 20% is better than nothing - did you create a branch that we could have a look at?

I can have a go at a C implementation next week, or as procrastination this week and see what I can do.
I will also play with concurrency in the call from matched-filter, or (probably better) create a more useful subfunc in find_peaks for concurrent work.

**Comment by d-chambers** [comment]
No, 20% is an estimated upper limit based on profiling the current function.

**Comment by yifanyin** [comment]
Concurrent find_peaks might help. Because it's done one by one on one core after the paralleled xcorr function. So what I used to do is just wrap it using pool -- same way as how _template_loop was wrapped. I haven't time the improvement though.

**Comment by calum-chamberlain** [comment]
So I have managed a 10x speed-up by writing a compiled C-func for similar datasets (with many peaks).  I will create a branch up here soon.  The speed-up only comes for datasets with many peaks in them, for datasets with few peaks, there is little to no speed-up, but in these cases the old way was also fast (<0.01s for my tests for 10,000 datapoints).  For tests with the same datapoints but many more peaks (\~3,000), the compiled version ran in 2.57s and the old version in 19.66s.

I should also be able to write a python multiprocessing function to wrap find_peaks in.

In doing this I did notice a bug - find-peaks wasn't handling minima in the way I expected it to - it was not including the minimum value, but a the highest value between the minima and the threshold (which could be a couple of samples out, and a value closer to the threshold).  I have fixed this in my branch and will include this in later versions. I don't think this will have had much impact on detections (apart from anticorrelated detections being a couple of samples out and having an incorrect detect_val associated with them), expect for detections separated from others by intervals close to trig_int.

A knock on problem could have been for lag-calc, but lag-calc doesn't use anticorrelated events, so this isn't a problem.

I need to run some more tests, and should really do real work today rather than fun procrastination, but @yifanyin if you want to test this stuff out it is the [findpeaks_optimisation](https://github.com/eqcorrscan/EQcorrscan/tree/findpeaks-optimisation) branch.

**Comment by d-chambers** [comment]
I have been playing around a little bit more with this as well. I made an implementation that takes the inner loop of decluster and uses numpy arrays instead. The speed-ups I have seen range from 3% to 1000% depending on the number of peaks passed to decluster. I will step back now that you have a C implementation.  

I did make a few more tests to ensure I understand how this function should behave. One such test, link below, iterates over the peaks and ensures the peak values are the max of the abs of the cc_array within the window specified by trig_index parameter. Curiously, this is not always the case. Could this be a manifestation of the bug you mentioned?

https://github.com/d-chambers/EQcorrscan/blob/c970fd62be3170cdeda1abacea6f69ca518bfab1/eqcorrscan/tests/find_peaks_test.py#L46

In the same branch you might find some of the tests I have already written useful. 

Also, in the future, is it better if I work off my fork of EQcorrscan or to put it in the EQcorrscan repo?

**Comment by calum-chamberlain** [comment]
1,000% is a pretty nice speed-up, and if you can do it with numpy rather than C then I would be keen to run that - how about I run some tests with your decluster and my C decluster and see what comes out faster? If we can avoid more C then I think that is a good thing. Yours is heaps cleaner than my old one was!

I'm fairly sure the issue you mention is a manifestation of the bug - this should only happen for peaks in the absolute values that are actually minima.

Those tests look better too - would be good to have them.

I think in general it is probably simpler if you work in branches in the EQcorrscan repo rather than a fork, we can then share work a bit more - but it is up to you.

**Comment by d-chambers** [comment]
Ya this was a really interesting problem. I am surprised with how hard it is to beat ndimage.find_objects based routine for the test case you already had written. I tried a few implementations, including using a numba jit'ed function but it ended up being slower. Reading the find_objects documentation, however, there might a problem in using the implementation for an extreme corner case, I am not sure if it is worth fixing if it slows things down considerably. Consider this:

```python
import numpy as np

from eqcorrscan.utils import find_peaks2_short

arr = np.array([1, .2, .2, .2, .2, 1, .2, .2, .2, .2, 1])

peaks = find_peaks2_short(arr, thresh=.2, trig_int=3)
```
I would expect peaks to be `[(1.0, 0), (1.0, 5), (1.0, 10)]` but because of how find_objects works it actually is `[(1.0, 0)]`, so in the case the threshold is really low it can connect larger blocks than one would expect from the trig_int parameter. Is this worth fixing if it causes the function to be several times slower? It will probably be rare this happens in the context of a correlation coefficients but not inconceivable. 

The branch I pointed you to is still in rough state so feel free to change anything if you even find any of it useful.

**Comment by calum-chamberlain** [comment]
Its really hard eh! I tried so many things when writing it, including fitting various functions, but they weren't reliable.

Curious edge-case, I will have a play and see what I think. It should be rare, but you are right, it is definitely possible.

Thanks, I will borrow things out of that branch to test 😸

**Comment by calum-chamberlain** [comment]
So one way I am testing to try and cope with the case you mention is to use the following rather than just taking the maximum in the window:
```python
    for peak_slice in peak_slices:
        # print('Width of peak='+str(peak_slice[0].stop-peak_slice[0].start)
        window = arr[peak_slice[0].start: peak_slice[0].stop]
        if peak_slice[0].stop - peak_slice[0].start > trig_int:
            peaks = decluster_compiled(peaks=list(zip(
                window, np.arange(peak_slice[0].start, peak_slice[0].stop))),
                trig_int=trig_int, debug=debug)
        else:
            peaks = [(window[np.argmax(abs(window))],
                      int(peak_slice[0].start + np.argmax(abs(window))))]
        initial_peaks.extend(peaks)
```

So, if the window is long then it finds peaks in it using the decluster method (whichever we end up using - I'm leaning towards yours at the moment, it is a similar speed to my compiled one and stays in Python).
Using that method gets the expected result, but at some cost. Initial tests look like it runs about .1s slower for the tests I have set up... Which is about 0.2% but I don't think they are good tests of it.
Opinions? Because it is an if-statement it slows down for all cases (through the if check), but only realllly slows down those case where something similar to you case happens?

**Comment by calum-chamberlain** [comment]
Hmm, I don't like that - then it finds lots of things that I wouldn't deem to be peaks. I think leaving it as it is is the best way to go.

e.g. in the test correlation function it finds 315 peaks, which are all above the threshold, and all separated by at-least trig-int, but I wouldn't personally deem them to be "peaks".

**Comment by d-chambers** [comment]
Agreed, this is probably a case where practicality beats purity but it is good to be aware of the limitation.

**Comment by calum-chamberlain** [comment]
An example with the test correlation. Red dots mark peaks found by peak-finding splitting windows longer than trig-int into multiple peaks (with a trig-int of 53 samples), and blue are the old ones.

Allowing multiple peaks to be extracted from the same window ends up putting "peaks" where there isn't a turning point (note, I have tried to implement finding peaks by calculating the derivative of the array, but it's not stable for noisy data).
![peak_finding](https://user-images.githubusercontent.com/10913395/30627847-e3d2786e-9e26-11e7-9d1b-eeb4f3a9b55b.png)

I think you are right - I should document this better and will probably use your example to show the limitations.

**Comment by d-chambers** [comment]
I made a slight change to decluster2 that improves things a bit by not making copies of one of the arrays for each peak:

```python
# sort peaks highest to lowest
peak_ind_list = []  # a list of indices that should be kept
valid_ind = np.ones(len(peaks), dtype=bool)
# get a sorted list of indices of the reduced arrays
new_inds = np.argsort(-np.abs(peaks))
# get a sorted list of the original indicies
sorted_peak_inds = index[new_inds]
# iterate, if no peak with a high value occurs within trig int save
for pnum, ind in enumerate(sorted_peak_inds):
    inds_bigger = np.abs(sorted_peak_inds[: pnum] - ind) < (trig_int + 1)
    if not np.any(inds_bigger & valid_ind[: pnum]):
        peak_ind_list.append(new_inds[pnum])
    else:
        valid_ind[pnum] = False

    # form a list of tuples with [(peak_value, peak_index), ...]
return [(peaks[x], index[x]) for x in np.sort(peak_ind_list)]
```

**Comment by calum-chamberlain** [comment]
Nice - I will put that in this afternoon, clean up my branch and run some tests then push up here and start a PR that we can comment and work on - because it will be branch up here you will be able to work on it as well if you want?
Or we can work from your branch if you can move your branch to be off the main repo rather than the fork?

**Comment by d-chambers** [comment]
Sounds good if you want to start the PR.

**Comment by calum-chamberlain** [comment]
PR opening in #167 including list of things to do.

**Comment by calum-chamberlain** [comment]
Because we have closed #167, I have moved the rest of the work outlined here to a new issue: #169 and will close this one.
